FFmpegInterop/UncompressedVideoSampleProvider.cpp

```diff
@@ -18,140 +18,111 @@
+			// Setup software scaler to convert the pixel format to NV12
+			m_swsContext.reset(sws_getContext(
+				m_codecContext->width,
+				m_codecContext->height,
+				m_codecContext->pix_fmt,
+				m_codecContext->width,
+				m_codecContext->height,
+				AV_PIX_FMT_NV12,
+				SWS_BICUBIC,
+				nullptr,
+				nullptr,
+				nullptr));
+			THROW_IF_NULL_ALLOC(m_swsContext);
+
+			THROW_HR_IF_FFMPEG_FAILED(av_image_fill_linesizes(m_lineSizes, AV_PIX_FMT_NV12, m_codecContext->width));
+
+			// Create a buffer pool
+			const int requiredBufferSize = av_image_get_buffer_size(AV_PIX_FMT_NV12, m_codecContext->width, m_codecContext->height, 1);
+			THROW_HR_IF_FFMPEG_FAILED(requiredBufferSize);
+
+			m_bufferPool.reset(av_buffer_pool_init(requiredBufferSize, nullptr));
+			THROW_IF_NULL_ALLOC(m_bufferPool);
 		}
 	}
 
-	if (SUCCEEDED(hr))
+	void UncompressedVideoSampleProvider::SetEncodingProperties(_Inout_ const IMediaEncodingProperties& encProp, _In_ bool setFormatUserData)
 	{
-		m_pAvFrame = av_frame_alloc();
-		if (m_pAvFrame == nullptr)
+		SampleProvider::SetEncodingProperties(encProp, setFormatUserData);
+
+		VideoEncodingProperties videoEncProp{ encProp.as<VideoEncodingProperties>() };
+
+		if (m_codecContext->framerate.num != 0 && m_codecContext->framerate.den != 0)
 		{
-			hr = E_OUTOFMEMORY;
+			MediaRatio frameRate{ videoEncProp.FrameRate() };
+			frameRate.Numerator(m_codecContext->framerate.num);
+			frameRate.Denominator(m_codecContext->framerate.den);
 		}
+
+		MediaPropertySet videoProp{ videoEncProp.Properties() };
+		videoProp.Insert(MF_MT_INTERLACE_MODE, PropertyValue::CreateUInt32(MFVideoInterlace_MixedInterlaceOrProgressive));
 	}
 
-	if (SUCCEEDED(hr))
+	tuple<IBuffer, int64_t, int64_t, map<GUID, IInspectable>> UncompressedVideoSampleProvider::GetSampleData()
 	{
-		if (av_image_alloc(m_rgVideoBufferData, m_rgVideoBufferLineSize, m_pAvCodecCtx->width, m_pAvCodecCtx->height, AV_PIX_FMT_NV12, 1) < 0)
+		// Get the next decoded sample
+		AVFrame_ptr frame{ GetFrame() };
+
+		// Get the sample buffer
+		IBuffer sampleBuf{ nullptr };
+		if (m_swsContext == nullptr)
 		{
-			hr = E_FAIL;
+			// Image is already in the desired output format
+			sampleBuf = make<FFmpegInteropBuffer>(frame->buf[0]);
 		}
-	}
+		else
+		{
+			// Scale the image to the desired output format
+			AVBufferRef_ptr bufferRef{ av_buffer_pool_get(m_bufferPool.get()) };
+			THROW_IF_NULL_ALLOC(bufferRef);
 
-	return hr;
-}
+			uint8_t* data[4]{ };
+			const int requiredBufferSize{ av_image_fill_pointers(data, AV_PIX_FMT_NV12, m_codecContext->height, bufferRef->data, m_lineSizes) };
+			THROW_HR_IF_FFMPEG_FAILED(requiredBufferSize);
+			THROW_HR_IF(MF_E_UNEXPECTED, requiredBufferSize != bufferRef->size);
 
-UncompressedVideoSampleProvider::~UncompressedVideoSampleProvider()
-{
-	if (m_pAvFrame)
-	{
-		av_frame_free(&m_pAvFrame);
-	}
+			THROW_HR_IF_FFMPEG_FAILED(sws_scale(m_swsContext.get(), frame->data, frame->linesize, 0, m_codecContext->height, data, m_lineSizes));
```

Comment:
>m_codecContext->height [](start = 92, length = 22)

Should use frame->height here (as mentioned in my comment above, the frame width / height can change dynamically).